File: crates/skald/skald-workflow/examples/with_observer.rs
//! Custom observer example attached with Workflow::with_observers.

use std::collections::HashMap;
use std::sync::Arc;
use std::sync::atomic::{AtomicU64, Ordering};

use async_trait::async_trait;
use skald_agent::Observer;
use skald_spec::{ProviderRequest, ProviderResponse};
use skald_workflow::{Workflow, WorkflowInput};
use wyrd_observe::OtelObserver;

mod common;
use common::{mock_registry, plan_prompt, write_prompt};

#[derive(Default)]
struct TokenCounter {
    calls: AtomicU64,
    tokens_in: AtomicU64,
    tokens_out: AtomicU64,
}

#[async_trait]
impl Observer for TokenCounter {
    async fn on_model_call(
        &self,
        _run_id: &str,
        _agent_id: &str,
        _iteration: u32,
        _provider: &str,
        _model: &str,
        _request: &ProviderRequest,
    ) {
        self.calls.fetch_add(1, Ordering::Relaxed);
    }

    async fn on_model_result(
        &self,
        _run_id: &str,
        _agent_id: &str,
        _iteration: u32,
        _finish_reason: &str,
        synthetic: bool,
        response: &ProviderResponse,
    ) {
        if synthetic {
            return;
        }
        if let ProviderResponse::OpenAiChatCompletion(response) = response {
            if let Some(usage) = &response.usage {
                self.tokens_in
                    .fetch_add(usage.prompt_tokens as u64, Ordering::Relaxed);
                self.tokens_out
                    .fetch_add(usage.completion_tokens as u64, Ordering::Relaxed);
            }
        }
    }
}

fn main() -> anyhow::Result<()> {
    let counter = Arc::new(TokenCounter::default());
    let planner = skald_agent::Agent::new(plan_prompt()).name("planner");
    let writer = skald_agent::Agent::new(write_prompt()).name("writer");
    let wf = Workflow::sequential("research", vec![planner, writer])?.with_observers(vec![
        Arc::new(OtelObserver::new()) as Arc<dyn Observer>,
        Arc::clone(&counter) as Arc<dyn Observer>,
    ]);
    let providers = mock_registry(&[
        r#"{"summary":"mock summary","steps":["read","write"]}"#,
        "Final observed brief",
    ]);
    let run = wyrd_runtime::runtime().block_on(wf.run_with(
        &providers,
        WorkflowInput::from(HashMap::from([(
            "topic".to_owned(),
            "the Rust borrow checker".to_owned(),
        )])),
    ))?;
    println!("model calls: {}", counter.calls.load(Ordering::Relaxed));
    println!("tokens in:   {}", counter.tokens_in.load(Ordering::Relaxed));
    println!(
        "tokens out:  {}",
        counter.tokens_out.load(Ordering::Relaxed)
    );
    println!("final:       {:?}", run.final_output);
    Ok(())
}

File: sdks/wyrd-sdk-python/examples/with_observer.py
"""Custom observer example attached to the workflow via observers=."""

import threading

from wyrd import Agent, Observer, OtelObserver, Prompt, Workflow


class TokenCounter(Observer):
    def __init__(self) -> None:
        self._lock = threading.Lock()
        self.calls = 0
        self.total_tokens_in = 0
        self.total_tokens_out = 0

    def on_model_call(self, run_id, agent_id, iteration, provider, model, request):
        with self._lock:
            self.calls += 1

    def on_model_result(self, run_id, agent_id, iteration, finish_reason, synthetic, response):
        if synthetic:
            return
        try:
            usage = response.openai().usage
        except Exception:
            return
        if usage:
            with self._lock:
                self.total_tokens_in += usage.prompt_tokens
                self.total_tokens_out += usage.completion_tokens


def main() -> None:
    counter = TokenCounter()
    planner = Agent(
        name="planner",
        prompt=Prompt(
            provider="mock",
            model="mock-model",
            messages=[r'{"summary":"mock summary","steps":["read","write"]}'],
            output={"summary": str, "steps": list[str]},
        ),
    )
    writer = Agent(
        name="writer",
        prompt=Prompt(provider="mock", model="mock-model", messages=["Brief: ${summary}"]),
    )
    wf = Workflow.sequential("research", planner, writer, observers=[OtelObserver(), counter])
    run = wf.run({"topic": "the Rust borrow checker"})
    print(f"model calls: {counter.calls}")
    print(f"tokens in:   {counter.total_tokens_in}")
    print(f"tokens out:  {counter.total_tokens_out}")
    print(f"final:       {run.final_output}")


if __name__ == "__main__":
    main()
